aptos-move/block-executor/src/scheduler_v2.rs

```diff
@@ -223,28 +242,152 @@ impl SchedulerV2 {
+    pub(crate) fn try_get_sequential_commit_hook(
+        &self,
+    ) -> Result<Option<(TxnIndex, Incarnation)>, PanicError> {
+        // Relaxed ordering due to armed lock acq-rel.
+        let next_to_commit_idx = self.next_to_commit_idx.load(Ordering::Relaxed) as usize;
+
+        assert!(next_to_commit_idx <= self.num_txns as usize);
+        if self.is_halted() || next_to_commit_idx == self.num_txns as usize {
+            // All sequential commit hooks are already dispatched.
+            return Ok(None);
+        }
+
+        if next_to_commit_idx > 0
+            && self.committed_marker[next_to_commit_idx - 1].load(Ordering::Relaxed) != COMMITTED
+        {
+            // Must wait for the previous hook to complete.
+            return Ok(None);
+        }
+
+        if self.txn_status[next_to_commit_idx as usize].is_executed() {
+            // All prior transactions are committed and the latest incarnation of the transaction
+            // at next_to_commit_idx has finished but has not been aborted. If any of its read was
+            // incorrect, it would have been invalidated by the respective transaction's last
+            // (committed) (re-)execution, and led to an abort in the corresponding finish execution
+            // (which, inductively, must occur before the transaction is committed). Hence, it
+            // must also be safe to commit the current transaction.
+
+            if self.committed_marker[next_to_commit_idx as usize]
+                .swap(PENDING_COMMIT_HOOK, Ordering::Relaxed)
+                != NOT_COMMITTED
+            {
+                return Err(code_invariant_error(format!(
+                    "Marking {} as PENDING_COMMIT_HOOK, but previous marker != NOT_COMMITTED",
+                    self.post_commit_processing_queue.len()
+                )));
+            }
+
+            // Transaction might be executed despite a stalled status, e.g. intentionally,
+            // due to having a high priority, or simply if it gets stalled after the
+            // execution task is picked up. Hence, a transaction can be committed while
+            // stalled, and in this case its aborted dependencies need to be unstalled.
+            let mut aborted_dependency_guard =
+                self.aborted_dependencies[next_to_commit_idx as usize].lock();
+            let mut propagation_queue = BTreeSet::new();
+            if aborted_dependency_guard.is_stalled {
+                aborted_dependency_guard.unstall(&self.txn_status, &mut propagation_queue)?;
+            }
+            self.propagate(propagation_queue)?;
+
+            // Increments cause cache invalidations, but finish_execution / try_commit
+            // performs only a relaxed read to check (see above).
+            let prev_idx = self
+                .next_to_commit_idx
+                .swap(next_to_commit_idx as u32 + 1, Ordering::Relaxed);
+            if prev_idx != next_to_commit_idx as u32 {
```

Comment:
Everything is under lock. Maybe this check is too much?

Reply:
The lock is acquired by the caller, because the pattern is acquire lock, call this & post-process in a loop, unlock - hence this can be useful if the caller doesn't properly hold the lock and use it incorrectly?
otherwise it could be an assert or not even check (store without swap).